createwindow/createwindow.go

```diff
@@ -0,0 +1,299 @@
+}
+
+func (s *Session) Newconn(displayString string) (*xproto.ScreenInfo, error) {
+	var err error
+	s.X, err = xgb.NewConnDisplay(displayString)
+	if err != nil {
+		return nil, err
+	}
+
+	setup := xproto.Setup(s.X)
+	screenInfo := setup.DefaultScreen(s.X)
+
+	return screenInfo, nil
+}
+
+func establishChromeConnection(programState *command.ProgramState, timeout int) (context.Context, error) {
+	wsURL, err := command.WsURL(programState, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("Could not connect to the chrome window. Encountered error %s", err.Error())
+	}
+
+	if wsURL == "" {
+		return nil, errors.New("must specify -devtools-ws-url")
+	}
+
+	allocatorContext, _ := chromedp.NewRemoteAllocator(context.Background(), wsURL)
+
+	ctx, _ := chromedp.NewContext(allocatorContext)
+
+	return ctx, nil
+}
+
+// ForceQuit closes all open windows
+func (s *Session) ForceQuit() {
+	programs := s.getChromeList()
+
+	fmt.Println("starting force quit")
+
+	for _, q := range programs {
+		if q.ToClose() == true {
+			q.Quit()
+		}
+	}
+	// Input Window is gracefully closed, closing closed window is okay.
+	s.InputWin.Quit()
+
+}
+
+func (s *Session) CreateXephyrWindow(layout Layout, display int, cmdErrorHandler func(p *command.ProgramState, err error) error) error {
+	if layout.h == 0 {
+		layout.h = windowHeight
+	}
+	if layout.w == 0 {
+		layout.w = windowWidth
+	}
+
+	displayString := fmt.Sprintf(":%d", display)
+	xephyr := command.ExternalCommand{
+		Path: "Xephyr",
+		Arg: []string{
+			displayString,
+			"-ac", "-screen",
+			fmt.Sprintf("%dx%d+%d+%d", layout.w, layout.h, layout.x, layout.y),
+			"-br",
+			"-reset",
+			"-no-host-grab",
+		},
+	}
+	programstate, err := command.ExecuteProgram(xephyr, cmdErrorHandler)
+	if err != nil {
+		return err
+	}
+	// the Xephyr window is not a chromewindow but maintained in the same list.
+	s.appendChromeList(ChromeWindow{programstate, nil})
+
+	for {
+		_, err = os.Stat("/tmp/.X11-unix/X" + strconv.Itoa(display))
+		if !os.IsNotExist(err) {
+			break
+		}
+	}
+	return nil
+}
+
+func DefaultXephyrSize() (height, width uint16) {
+	height, width = 900, 1600
+
+	X, err := xgb.NewConn()
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	if err := randr.Init(X); err != nil {
+		log.Println(err)
+		return
+	}
+
+	screens, err := randr.GetScreenResourcesCurrent(X, xproto.Setup(X).DefaultScreen(X).Root).Reply()
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	crtc, err := randr.GetCrtcInfo(X, screens.Crtcs[0], xproto.TimeCurrentTime).Reply()
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	return uint16(0.8 * float64(crtc.Height)), uint16(0.8 * float64(crtc.Width))
+}
+
+// Setup opens all windows and establishes connection with the x server
+func Setup(n int) (*Session, error) {
+	debuggingport := 9222
+	var displayString string
+	var session Session
+
+	// calls ForceQuit in case ChromeWindows are closed
+	cmdErrorHandler := func(p *command.ProgramState, err error) error {
```

Comment:
p *command.ProgramState is unused? Remove it from the callback definition then

And then you can use the exact same callback for UnmapNotify with err = nil (they both basically do the same thing?)

You should also typedef the func to something like ExitCallbackFn so that this is easier to read